### ISSUE TITLE
Flutter crashes on bad assets paths in pubspec.yaml

### ISSUE THREAD

**Issue by perlatus** [flutter/flutter#12460]
# tl;dr
An asset path that doesn't exist causes flutter to crash with an obscure error message. It's not terribly hard to read through the log and determine the cause of the crash, but the current state is rather unfriendly.

Potentially related to #3937.

# Steps to reproduce
1. `flutter create crash`
2. Uncomment  lines in pubspec.yaml so it's like below:

```
  # To add assets to your application, add an assets section, like this:
  assets:
    - images/a_dot_burr.jpeg
  #  - images/a_dot_ham.jpeg
```
3. `cd crash; flutter run`

# Result
```
~/tmp/crash> flutter run
Running "flutter packages get" in crash...            0.5s
Launching lib/main.dart on XT830C in debug mode...
Initializing gradle...                                0.6s
Resolving dependencies...                             0.9s
Running 'gradlew assembleDebug'...                        

Oops; flutter has exited unexpectedly.
Crash report written to /home/vczf/tmp/crash/flutter_01.log;
please let us know at https://github.com/flutter/flutter/issues.

FAILURE: Build failed with an exception.

* Where:
Script '/data/code/clone/flutter/packages/flutter_tools/gradle/flutter.gradle' line: 350

* What went wrong:
Execution failed for task ':app:flutterBuildDebug'.
> Process 'command '/data/code/clone/flutter/bin/flutter'' finished with non-zero exit value 1

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.
Gradle build failed: 1
Exception: flutter exited with 1
Traceback:
  [interactive], line 1:
    flutter run
```

## flutter_01.log

Flutter crash report; please file at https://github.com/flutter/flutter/issues.

## command

flutter build flx --suppress-analytics --target /home/vczf/tmp/crash/lib/main.dart --output-file /home/vczf/tmp/crash/build/app/intermediates/flutter/debug/app.flx --snapshot /home/vczf/tmp/crash/build/app/intermediates/flutter/debug/snapshot_blob.bin --depfile /home/vczf/tmp/crash/build/app/intermediates/flutter/debug/snapshot_blob.bin.d --working-dir /home/vczf/tmp/crash/build/app/intermediates/flutter/debug/flx

## exception

FileSystemException: FileSystemException: Directory listing failed, path = '/home/vczf/tmp/crash/images/' (OS Error: No such file or directory, errno = 2)

```
#0      _Directory._fillWithDirectoryListing (dart:io-patch/directory_patch.dart:28)
#1      _Directory.listSync (dart:io/directory_impl.dart:202)
#2      _LocalFileSystemEntity&ForwardingDirectory.listSync (package:file/src/forwarding/forwarding_directory.dart:43)
#3      _AssetDirectoryCache.variantsFor (package:flutter_tools/src/asset.dart:453)
#4      _parseAssets (package:flutter_tools/src/asset.dart:496)
#5      AssetBundle.build (package:flutter_tools/src/asset.dart:101)
<asynchronous suspension>
#6      assemble (package:flutter_tools/src/flx.dart:110)
<asynchronous suspension>
#7      build (package:flutter_tools/src/flx.dart:79)
<asynchronous suspension>
#8      BuildFlxCommand.runCommand (package:flutter_tools/src/commands/build_flx.dart:44)
<asynchronous suspension>
#9      FlutterCommand.verifyThenRunCommand (package:flutter_tools/src/runner/flutter_command.dart:253)
<asynchronous suspension>
#10     FlutterCommand.run (package:flutter_tools/src/runner/flutter_command.dart:193)
<asynchronous suspension>
#11     CommandRunner.runCommand (package:args/command_runner.dart:194)
<asynchronous suspension>
#12     FlutterCommandRunner.runCommand (package:flutter_tools/src/runner/flutter_command_runner.dart:279)
<asynchronous suspension>
#13     CommandRunner.run.<anonymous closure> (package:args/command_runner.dart:109)
#14     new Future.sync (dart:async/future.dart:222)
#15     CommandRunner.run (package:args/command_runner.dart:109)
#16     FlutterCommandRunner.run (package:flutter_tools/src/runner/flutter_command_runner.dart:162)
#17     run.<anonymous closure> (package:flutter_tools/executable.dart:143)
<asynchronous suspension>
#18     AppContext._run (package:flutter_tools/src/base/context.dart:76)
<asynchronous suspension>
#19     AppContext.runInZone.<anonymous closure> (package:flutter_tools/src/base/context.dart:66)
#20     _rootRun (dart:async/zone.dart:1120)
#21     _CustomZone.run (dart:async/zone.dart:1001)
#22     runZoned (dart:async/zone.dart:1467)
#23     AppContext.runInZone (package:flutter_tools/src/base/context.dart:65)
#24     run (package:flutter_tools/executable.dart:114)
<asynchronous suspension>
#25     main (package:flutter_tools/executable.dart:67)
<asynchronous suspension>
#26     main (file:///data/code/clone/flutter/packages/flutter_tools/bin/flutter_tools.dart:16)
#27     _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:263)
#28     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:151)
```

## flutter doctor

```
[✓] Flutter (on Linux, locale en_US.UTF-8, channel master)
    • Flutter at /data/code/clone/flutter
    • Framework revision 4e48a737eb (2 days ago), 2017-10-06 19:24:34 -0700
    • Engine revision 459f722b86
    • Tools Dart version 1.25.0-dev.11.0

[✓] Android toolchain - develop for Android devices (Android SDK 26.0.1)
    • Android SDK at /home/vczf/opt/android-sdk
    • Platform android-26, build-tools 26.0.1
    • Java binary at: /home/vczf/opt/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] Android Studio (version 2.3)
    • Android Studio at /home/vczf/opt/android-studio
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] Android Studio
    • Android Studio at /opt/android-studio
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] Connected devices
    • XT830C • ZX1B72G4V6 • android-arm • Android 4.4.4 (API 19)

```
